Addon-docs/Angular: Inline rendering support with angular-elements

ThibaudAV opened the pull request:
Issue: #8153 #12264 #13480


Another PR : https://github.com/storybookjs/storybook/pull/11339

## TODO / Question :

- [x] Check the optional dependencies on angular element works correctly on a project that is not in monorepo.
- [x] Add / correct the doc
- [ ] ~~Optimization with change of props (avoids refresh the whole component if props change)~~ (if necessary in another PR)
- [x] Add a more complete example in angular-cli?
- [x] Should we activate inlineStories by default? 
- [ ] Add tests

## What I did

Uses angular element to transform story component into customElements (web-component). 
The custom element is then added in React. 


Add optional `@storybook/angular` dependency to `@storybook/addon-docs`.
Add optional `@angular/elements` & `@webcomponents/custom-elements` dependencies in `@storybook/angular`. 
Only one file (`ElementRendererService `) uses these dependencies, so if it is not called these depeandace still optional. It seems to me 🤔 
:not-sure:

User has to specify these dependencies if he uses `prepareForInline` 

I tried to add them in the addon-docs but it adds more angular dependency. I figure it's better if they're all in `@storybook/angular`? 
What do you think about it?


## How it work

Add : `yarn add -D @angular/elements @webcomponents/custom-elements`
Then update `.storybook/preview.js`:

```js
import { addParameters } from '@storybook/angular';
import { prepareForInline } from '@storybook/addon-docs/angular/inline';
addParameters({
  docs: {
    inlineStories: true,
    prepareForInline,
  },
});
```

A special rendering is made by `prepareForInline`:
- is called for each story 
- `prepareForInline` use `ElementRendererService` from `@storybook/angular`
- `ElementRendererService` use singleton instance of angular platform from `RendererService `
- `ElementRendererService` generate custom `ElementsModule` from `storyInput` ("main component" is generated in the same way as the "classic render")
- `ElementsModule` is bootstrap by angular in `<div id="root">`
- `ElementsModule` use `createCustomElement` from `@angular/elements` to extract the main component (the one of the story) and return them to `prepareForInline` as customElement
- `prepareForInline` define in `customElements` new customElement

Note: as it is not possible to delete defined customElements. An "increment" allows to make each new generation unique.

## How to test

- Is this testable with Jest or Chromatic screenshots? 
- Does this need a new example in the kitchen sink apps?
- Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

ThibaudAV committed:
fix(angular): add right Parameters type in StoryContext type

shilman reviewed (commented):
This is great and will make a lot of people happy. I don't understand enough about the Angular/Web components story to comment on the high-level approach, but I believe @stupidawesome suggested this in a previous conversation with @kroeder and me back when we were discussing Ivy rendering.

@kroeder @stupidawesome @costalvaro @petermikitsh @manuelmeister can you please take a look at this PR and give feedback?

kroeder commented on app/angular/src/client/preview/angular-beta/StorybookWrapperComponent.ts:
```diff
@@ -0,0 +1,137 @@
+  class StorybookWrapperComponent implements AfterViewInit, OnDestroy {
+    private storyPropsSubscription: Subscription;
+
+    @ViewChild(storyComponent ?? '', { static: true }) storyComponentElementRef: ElementRef;
```
Why do you need this? What can I expect from a `@ViewChild('')`?

ThibaudAV replied:
Before there were 2 wrappers (one for component and one for template). 
There is now only 1 for both. And in case of a story without component the ViewChild are "disabled" with `''` 

This change has related to this open PR https://github.com/storybookjs/storybook/pull/13507
(I'm based on this one)

kroeder reviewed (commented):
Nice work so far 🙂 

Your plan is to solve the issues with Storybooks docs mode, right? 

> I tried to add them in the addon-docs but it adds more angular dependency. I figure it's better if they're all in @storybook/angular?
What do you think about it?

@shilman Something you might be able to give feedback to.

ThibaudAV commented:
thanks :) 

> Your plan is to solve the issues with Storybooks docs mode, right?

yes to propose an "inline" alternative to avoid iframe (adapts to the size of the component, changes in line with docs. can change theme with toolbar dynamically, ...)

shilman commented:
cc @mandarini

ThibaudAV committed:
feat(angular): add inline stories support  in the docs addon

The configuration has to be set up in each angular project to avoid forcing dependencies to `@angular/elements` ans `@webcomponents/custom-elements`

ThibaudAV commented:
Ty it here https://deploy-preview-13525--storybookjs.netlify.app/

ThibaudAV committed:
refactor(angular): move docs stories in Addon/Docs

ThibaudAV committed:
chore(angular): add more examples for docs addon

ThibaudAV committed:
chore(angular): add theme-decorator example

ThibaudAV commented:
I'm not sure how to test. :/ 
if you have ideas, I'm interested :) 
I figure it can be done with chromatic when / if this feature is used by default. 🤷‍♂️

shilman commented on examples/angular-cli/src/stories/addons/docs/doc-button/doc-button.stories.ts:
```diff
@@ -1,9 +1,8 @@
 
 export default {
-  title: 'DocButton',
+  title: 'Addons/Docs/DocButton',
```
All the other `angular-cli` examples are in `Addon/x`. I think "addons" is better, but I suggest making the switch in a separate PR and updating all the other examples at the same time.

shilman commented on examples/angular-cli/src/stories/addons/docs/doc-button/doc-button.stories.ts:
```diff
@@ -1,9 +1,8 @@
 
 export default {
-  title: 'DocButton',
+  title: 'Addons/Docs/DocButton',
```
All the other `angular-cli` examples are in `Addon/x`. I think "addons" is better, but I suggest making the switch in a separate PR and updating all the other examples at the same time.

ThibaudAV replied:
aaa you have unmasked me
I would like to rework all stories example. group them, sort them,...,  in order to see quickly their goals
I will do it in another PR

shilman reviewed (changes requested):
Working great! One small change.

ThibaudAV commented:
I do a final manual test :
-> New project with storyshots addon and without angular element 

And then it will seem good to me

shilman commented:
Let me know when you're ready. Super excited about this!!!

ThibaudAV commented:
@shilman My test was useful and I had to make a change : 

add a specific file so that the whole code branch (with element import) is not used if the user does not use `prepareForInline`
https://github.com/storybookjs/storybook/pull/13525/commits/94230a2e3c1f4247ad34ec4f6c21eac2ae53e646#diff-61c1e0294b5dcd012e60f3b1aa8cb382baee5a19876bd82b22394716b90a4e46R1

shilman commented:
@ThibaudAV are we good to go now after the tests pass?

ThibaudAV commented:
yes it's ok for me

shilman commented:
There's an e2e failure but it's related to a different PR that I merged and @phated is following up on that. So I'm merging this since everything else looks great. ❤️

danielleroux commented:
@ThibaudAV Hi iam testing this demo feature in our component library (Angular 10) but iam discovered a strange behavior in my project. I have to compile my library with the ngcc compiler after each build, otherwise i get the following error message:

`
zone.js:690 Unhandled Promise rejection: Cannot read property 'id' of undefined ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot read property 'id' of undefined
    at registerNgModuleType (core.umd.js:24829)
    at core.umd.js:24840
    at Array.forEach (<anonymous>)
    at registerNgModuleType (core.umd.js:24840)
    at new NgModuleFactory (core.umd.js:24936)
    at compileNgModuleFactory__POST_R3__ (core.umd.js:28472)
    at PlatformRef.push../node_modules/@angular/core/bundles/core.umd.js.PlatformRef.bootstrapModule (core.umd.js:28716)
    at ElementRendererService.<anonymous> (ElementRendererService.js:89)
    at step (ElementRendererService.js:53)
    at Object.next (ElementRendererService.js:34) TypeError: Cannot read property 'id' of undefined
    at registerNgModuleType (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:76527:31)
    at http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:76538:51
    at Array.forEach (<anonymous>)
    at registerNgModuleType (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:76538:21)
    at new NgModuleFactory (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:76634:17)
    at compileNgModuleFactory__POST_R3__ (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:80170:29)
    at PlatformRef.push../node_modules/@angular/core/bundles/core.umd.js.PlatformRef.bootstrapModule (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:80414:20)
    at ElementRendererService.<anonymous> (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:140803:26)
    at step (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:140767:23)
    at Object.next (http://localhost:6006/vendors~main.331aca382febe703f812.bundle.js:140748:53)
`
